src/huggingface_hub/_commit_api.py

```diff
@@ -449,9 +483,61 @@ def fetch_upload_modes(
     return upload_modes
 
 
+@validate_hf_hub_args
+def fetch_files_to_copy(
+    copies: Iterable[CommitOperationCopy],
+    repo_type: str,
+    repo_id: str,
+    token: Optional[str],
+    revision: str,
+    endpoint: Optional[str] = None,
+) -> Dict[Tuple[str, Optional[str]], "RepoFile"]:
+    """
+    Requests the Hub "preupload" endpoint to determine whether each input file
+    should be uploaded as a regular git blob or as git LFS blob.
+
+    Args:
+        copies (`Iterable` of :class:`CommitOperationCopy`):
+            Iterable of :class:`CommitOperationCopy` describing the files to
+            copy on the Hub.
+        repo_type (`str`):
+            Type of the repo to upload to: `"model"`, `"dataset"` or `"space"`.
+        repo_id (`str`):
+            A namespace (user or an organization) and a repo name separated
+            by a `/`.
+        token (`str`, *optional*):
+            An authentication token ( See https://huggingface.co/settings/tokens )
+        revision (`str`):
+            The git revision to upload the files to. Can be any valid git revision.
+
+    Returns: `Dict[Tuple[str, Optional[str]], RepoFile]]`
+        Key is the file path and revision of the file to copy, value is the repo file.
+
+    Raises:
+        [`~utils.HfHubHTTPError`]
+            If the Hub API returned an error.
+        [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError)
+            If the Hub API response is improperly formatted.
+    """
+    from .hf_api import HfApi
+
+    hf_api = HfApi(endpoint=endpoint, token=token)
+    files_to_copy = {}
+    for src_revision, operations in groupby(copies, key=lambda op: op.src_revision):
+        operations = list(operations)  # type: ignore
+        paths = [op.src_path_in_repo for op in operations]
+        src_repo_files = hf_api.list_files_info(
```

Comment:
`list_files_info` can be a bit misleading but  basically you cannot assume that the returned files will be in the same order as the input paths + you cannot even assume that the list will be the same size. In particular:
- if a path is duplicated (i.e. user is copying a file multiple times), the server will return only one RepoFile instance
- if a path is missing on the repo, the server will silently ignore it

So what I would advice to do is:
- take as input a `List[CommitOperationCopy]` instead of `Iterable`
- do a first pass with `for src_revision, operations in groupby(copies, key=lambda op: op.src_revision):`
  - `list_files_info`
  - if a returned RepoFile has `.lfs = None` => raise an issue
  - add to `files_to_copy` using `RepoFile.rfilename`, `src_revision` and `RepoFile.lfs["sha256"]`
- do a second pass on `copies`
  - if `op.src_path_in_repo`/`op.src_revision` pair is not found in the sha256 dictionary => raise the EntryNotFound error

_(disclaimer: I haven't tested the above logic)_

Reply:
You can iterate on a list twice. On an iterable it's not guaranteed

Reply:
Now I only iterate on it once using groupby so we're good